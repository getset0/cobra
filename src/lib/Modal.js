import styled from "styled-components";
import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTimes } from "@fortawesome/free-solid-svg-icons";
import H2 from "./Typography/H2";

const ModalWrapper = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  width: 100%;
  height: 100%;
  background: rgba(0, 0, 0, .75);
  display: flex;
  justify-content: center;
  align-items: center;
  z-index: 990;
`;

const ModalContent = styled.div`
  position: relative;
  width: 475px;
  height: 400px;
  background: #fff;
  padding: 10px;
  z-index: 999;
`;

const TitleWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  padding: 5px;
`;

const CustomModalContent = ({ children, id, title, handleCloseModal }) => (
  <ModalContent id={id}>
    <TitleWrapper>
      <H2>{title}</H2>
      <span style={{ fontSize: "18px", cursor: "pointer" }}>
        <FontAwesomeIcon
          icon={faTimes}
          onClick={handleCloseModal}
        />
      </span>
    </TitleWrapper>
    <div>
      {children}
    </div>
  </ModalContent>
);

class CustomModal extends React.Component {
  handleOverlayClick = (e) => {
    e.stopPropagation();
    e.target.id === "modal-overlay" && this.props.handleCloseModal();
  }
  render() {
    const { isOpen, children, title, handleCloseModal } = this.props;
    return isOpen ? (
      <ModalWrapper onClick={this.handleOverlayClick} id="modal-overlay">
        <CustomModalContent
          id="modal-content"
          title={title}
          handleCloseModal={handleCloseModal}
        >
          {children}
        </CustomModalContent>
      </ModalWrapper>
    ) : null;
  }
}


export default CustomModal;
