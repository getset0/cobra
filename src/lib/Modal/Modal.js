import styled from "styled-components";
import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTimes } from "@fortawesome/free-solid-svg-icons";
import H2 from "../Typography/H2";
import PropTypes from "prop-types";

export const ModalWrapper = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  width: 100%;
  height: 100%;
  background: rgba(0, 0, 0, 0.75);
  display: flex;
  justify-content: center;
  align-items: center;
  z-index: 990;
`;

export const ModalContent = styled.div`
  position: relative;
  width: 475px;
  height: 400px;
  background: #fff;
  padding: 10px;
  z-index: 999;
`;

export const TitleWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  padding: 5px;
`;

export const CustomModalContent = ({
  children,
  id,
  title,
  handleCloseModal
}) => (
  <ModalContent id={id}>
    <TitleWrapper>
      <H2>{title}</H2>
      <span style={{ fontSize: "18px", cursor: "pointer" }}>
        <FontAwesomeIcon icon={faTimes} onClick={handleCloseModal} />
      </span>
    </TitleWrapper>
    <div>{children}</div>
  </ModalContent>
);

class CustomModal extends React.PureComponent {
  handleOverlayClick = e => {
    if (e) {
      e.stopPropagation();
      e.target.id === "modal-overlay" && this.props.handleCloseModal();
    } else {
      this.props.handleCloseModal();
    }
  };
  render() {
    const { isOpen, children, title, handleCloseModal } = this.props;
    return isOpen ? (
      <ModalWrapper onClick={this.handleOverlayClick} id="modal-overlay">
        <CustomModalContent
          id="modal-content"
          title={title}
          handleCloseModal={handleCloseModal}
        >
          {children}
        </CustomModalContent>
      </ModalWrapper>
    ) : null;
  }
}

CustomModal.propTypes = {
  children: PropTypes.any.isRequired,
  isOpen: PropTypes.bool.isRequired,
  handleCloseModal: PropTypes.func.isRequired,
  title: PropTypes.string
};

CustomModal.defaultProps = {
  isOpen: false,
  title: "Modal title"
};

export default CustomModal;
