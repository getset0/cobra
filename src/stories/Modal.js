import React from "react";
import { storiesOf } from "@storybook/react";
import { withInfo } from "@storybook/addon-info";
import { withBackgrounds } from "@storybook/addon-backgrounds";
import { backgroundOptions } from "./util";
import { Modal, Button } from "../lib";

class TestModal extends React.Component {
    state = {
      isModalOpen: false
    };
    handleOpenModal = () => this.setState({ isModalOpen: true });
    handleCloseModal = () => this.setState({ isModalOpen: false });
    render() {
      return (
        <React.Fragment>
          <Button onClick={this.handleOpenModal}>Open Modal</Button>
          <Modal
            isOpen={this.state.isModalOpen}
            handleCloseModal={this.handleCloseModal}
            title="I am a modal title"
          >
            I am a modal content!
          </Modal>
        </React.Fragment>
      );
    }
}

storiesOf("Modal", module)
  .addDecorator(withInfo)
  .addDecorator(withBackgrounds(backgroundOptions))
  .add("Modal", () => <TestModal />);
