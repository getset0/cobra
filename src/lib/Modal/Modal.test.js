import React from "react";
import Modal, { ModalWrapper, CustomModalContent } from "./Modal";
import "jest-styled-components";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

it("renders without crash", () => {
  shallow(
    <Modal
      isOpen={true}
      handleCloseModal={jest.fn(x => x)}
      title="I am a modal title"
    >
    I am a modal content!
    </Modal>
  );
});

it("renders Tabs correctly", () => {
  const wrapper = shallow(
    <Modal
      isOpen={true}
      handleCloseModal={jest.fn(x => x)}
      title="I am a modal title"
    >
    I am a modal content!
    </Modal>
  );
  const tree = toJson(wrapper);
  expect(tree).toMatchSnapshot();
});

it("test Modal close", () => {
  const mockState = {
    isOpen: true
  };
  const mockStateChange = jest.fn(() => {
    mockState.isOpen = false;
  });
  const wrapper = shallow(
    <Modal
      isOpen={mockState.isOpen}
      handleCloseModal={mockStateChange}
      title="I am a modal title"
    >
    I am a modal content!
    </Modal>
  );

  // has ModalWrapper Component
  const overlay = wrapper.find(ModalWrapper);
  expect(overlay).toHaveLength(1);

  // has ModalContent Component
  const content = wrapper.find(CustomModalContent);
  expect(content).toHaveLength(1);

  // click on overlay
  overlay.prop("onClick")();

  // expect to have called mock 1 time
  expect(mockStateChange.mock.calls).toHaveLength(1);

  // ModalWrapper Component undefined
  const overlay2 = wrapper.find(ModalWrapper);
  expect(overlay2).toBeUndefined;

  // ModalContent Component undefined
  const content2 = wrapper.find(CustomModalContent);
  expect(content2).toBeUndefined;
});
