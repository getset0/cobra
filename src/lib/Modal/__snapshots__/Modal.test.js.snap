// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Tabs correctly 1`] = `
<styled.div
  id="modal-overlay"
  onClick={[Function]}
>
  <Component
    handleCloseModal={[Function]}
    id="modal-content"
    title="I am a modal title"
  >
    I am a modal content!
  </Component>
</styled.div>
`;
